csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js

```diff
@@ -0,0 +1,207 @@
+    } else if (timeUnits == 'centuries') {
+      var format = ngettext('1 century', '%s centuries', result);
+    }
+    var output = interpolate(format, [result]);
+    $('#output').val(output);
+  }
+});
+
+
+/** Calculates the time the algorithm will take to finish
+ *  executing in the selected time units */
+function calculateTimeTaken(complexity, resultForm, n, speed, processors, timeUnits) {
+  // can only have whole integers for number of items
+  n = Math.round(n);
+  $('#n-items').val(n);
+  processors = Mathjs.bignumber(Math.round(processors));
+  $('#processors').val(processors);
+  speed = Mathjs.bignumber(speed);
+  var steps;
+  
+  if (complexity == 'log') {
+    steps = Math.ceil(Mathjs.log(n, 2));
+  } else if (complexity == 'linear') {
+    steps = Mathjs.bignumber(n);
+  } else if (complexity == 'nlog') {
+    // because log(1) == 0
+    if (n == 1) {
+      steps = 1;
+    } else {
+      steps = Math.ceil(n * Mathjs.log(n, 2));
+    }
+  } else if (complexity == 'squared') {
+    steps = Mathjs.bignumber(Math.pow(n, 2));
+  } else if (complexity == 'cubed') {
+    steps = Mathjs.bignumber(Math.pow(n, 3));
+  } else if (complexity == 'fourth-power') {
+    steps = Mathjs.bignumber(Math.pow(n, 4));
+  } else if (complexity == 'sixth-power') {
+    steps = Mathjs.bignumber(Math.pow(n, 6));
+  } else if (complexity == 'exponential') {
+    steps = Mathjs.bignumber(Math.pow(2, n));
+  } else if (complexity == 'factorial') {
+    steps = Mathjs.factorial(n);
+  }
+
+  var denominator = Mathjs.multiply(speed, processors);
+  var timeTaken = Mathjs.divide(steps, denominator);
+
+  var timeScale = Mathjs.bignumber(TIME_SCALERS[timeUnits])
+  timeTaken = Mathjs.multiply(timeTaken, timeScale);
+
+  if (resultForm == 'scientific') {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'exponential', precision: 3});
+  } else {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'fixed', precision: 2});
+      // If there are no significant decimal places don't show any decimal places (non scientific mode only)
+    if (timeTaken % 1 == 0) {
+      timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'fixed', precision: 0});
+    }
+  }
+
+  return timeTaken;
+}
+
+
+/** Checks if the user input is valid */
+function inputIsValid(n, speed, processors) {
+  isValid = true;
+
+  // validation for number of items
+  nItems = $('#n-items');
+  nItemsErrorMsg = $('#n-items-input-error');
+  if (n < 1 || n > 1000 || isNaN(n)) {
+    nItems.addClass('is-invalid');
+    nItemsErrorMsg.removeClass('d-none');
+    isValid = false;
+    $('#output').val('');
+  } else if ((1 <= n <= 1000) && !nItemsErrorMsg.hasClass('d-none')) {
+    nItems.removeClass('is-invalid');
+    nItemsErrorMsg.addClass('d-none');
+  }
+
+  // validation for speed
+  speedInput = $('#speed');
+  speedErrorMsg = $('#speed-input-error');
+  if (speed <= 0 || speed > 10000000 || isNaN(speed)) {
```

Comment:
The HTML is 1 million but this is 10 million. Is that because of the multiply by ten thing? If so perhaps note it in a comment

Reply:
No that is city-trip. From our discussion of this I think it should be 1 million